Make Polygon2D emit polygon_changed when changed

poke1024 opened the pull request:
This PR solves the problem of detecting when a Polygon2D's vertices change.

Here's what I'm trying to achieve: get a signal each time a Polygon2D changed (esp. in the editor), so I can adjust other items that depend on the polygon.

Example: I use the following to keep the green ball attached to the first polygon vertex at all times:

```
tool
extends Polygon2D

func _enter_tree():
	connect("item_rect_changed", self, "on_item_rect_changed")
	
func on_item_rect_changed():
	get_node("../circle").set_position(transform * get_polygon()[0])
```

<img width="220" alt="bildschirmfoto 2017-08-26 um 17 22 24" src="https://user-images.githubusercontent.com/11859538/29742632-2a0af6a6-8a83-11e7-9d2c-0014720bd9d6.png">

<img width="474" alt="bildschirmfoto 2017-08-26 um 17 19 52" src="https://user-images.githubusercontent.com/11859538/29742626-02cb1c42-8a83-11e7-8bd7-e5bc1826079e.png">

With the current Polygon2D implementation, there seems no way to achieve this, as changes in vertices do not seem to generate any signals. The least I would expect is to get a `item_rect_changed`. This is what this PR implements.

Obviously this is not an optimal solution, since changes to vertices do not actually imply item rect changes. So would introducing a new signal type "vertices_changes" be the better alternative?

**EDIT** This PR no longer uses `item_rect_changed`, it now uses dedicated signal(s) and thus solves this last problem. ***END EDIT***

groud commented:
Not sure if this is the right way to solve this, as a vertice could change while not impacting the rect. I think this deserves a dedicated signal.

groud commented:
So I I agree with your last paragraph ^^

poke1024 commented:
I've removed `item_rect_changed` for now (something for a different PR as it gets quite complicated due to the item rect caching in `Polygon2D`) and added two new signals `polygon_changed` and `offset_changed`.

Changes of the accompanying transform (i.e. global vertex positions) can be additionally monitored via notifications (NOTIFICATION_TRANSFORM_CHANGED).

In order to monitor a polygon's global vertex positions, the script from above would now look like this:

```
tool
extends Polygon2D

func _enter_tree():
	connect("polygon_changed", self, "on_polygon_changed")
	connect("offset_changed", self, "on_polygon_changed")
	set_notify_transform(true)
	
func _notification(what):
	if what == NOTIFICATION_TRANSFORM_CHANGED:
		on_polygon_changed()
	
func on_polygon_changed():
	get_node("../circle").set_position(transform * (get_polygon()[0] + get_offset()))
```

poke1024 commented:
There are two things I keep wondering about:

(1) should there only be one signal that combines `polygon_changed` and `offset_changed`? I'm not sure what the general guideline in Godot is here. I think that mapping set/get handlers here is clean and I've seen `offset_changed` signals in other parts of the code (`graph_node` I believe).
(2) should there be a check such that `polygon_changed` only gets emitted if there's an actual change in `set_polygon`? If yes, it would involve checking all vertex coordinates, which an unnecessary overhead, if no signal is attached. One could fix this by adding some`Object::has_emit_signal` to check if an emit signal is connected, so that this expensive "is changed" check only happens for if signals are attached. A similar problem occurs if one wants to add `item_rect_changed` to Polygon2D, as Polygon2D caches its item rect/defers the calculation.

I'm fine with not doing neither (1) or (2), I just want to open this for discussion.

groud commented:
I don't think this offset_changed signal is very useful. In my opinion a coherent behaviour would be:
- One signal emitted when the polygon rect changes. Triggered when:
  - The polygon offset is modified,
  - A vertex is dragged and modifies the rect,
- One "polygon_changed" signal emitted every time the rect changes, but also when moving a single vertex.

poke1024 commented:
@groud This now should work as sketched out by you. Now using `get_signal_connection_list`.

akien-mga commented on scene/2d/polygon_2d.cpp:
```diff
@@ -382,6 +407,9 @@ void Polygon2D::_bind_methods() {
 	ADD_PROPERTY(PropertyInfo(Variant::REAL, "invert_border", PROPERTY_HINT_RANGE, "0.1,16384,0.1"), "set_invert_border", "get_invert_border");
+
+	ADD_SIGNAL(MethodInfo("polygon_changed"));
+	ADD_SIGNAL(MethodInfo("offset_changed"));
```
You're not using that signal in the end.

akien-mga commented:
I think this is growing pretty complex for a quite simple use case. If you need a signal emitted when changing the polygon, just emit `polygon_changed` and it should be fine, no?

poke1024 commented:
@akien-mga Yes, that's a good solution. Sorry for all the confusion. Changed now.

groud commented:
This PR looks good to me now :)

poke1024 commented:
I'm now wondering if there should even be a dedicated polygon_changed signal. I've seen `navigation_poly` uses a generic changed signal (`emit_signal(CoreStringNames::get_singleton()->changed)`), so maybe the cleaner and more complete solution would be to signal `change` whenever the polygon, the offset, the colors or the texture change?

poke1024 commented:
So, the use `CoreStringNames::get_singleton()->changed` seems to be `Resource`-specific, so a dedicated signal seems to make sense here, kind of. Updated now to emit a signal whenever any property changes. This should handle pretty every possible use case now.

poke1024 commented:
Here's a little demo on why this PR would be so useful. Mirrors a poly's form (and color) live in the editor:

![demo](https://user-images.githubusercontent.com/11859538/30127128-6990f140-933e-11e7-9733-c41373a9df16.gif)

Script used:

```
tool
extends Polygon2D

func on_changed():
	var p1 = get_polygon()
	var p2 = PoolVector2Array()
	for p in p1:
		p2.append(Vector2(-p.x, p.y))
	var b = get_node("../B")
	b.set_polygon(p2)
	b.set_color(get_color())

func _enter_tree():
	connect("polygon_changed", self, "on_changed")
```

akien-mga commented:
Sorry for the long delay. To me the current changes look good, I'd just like them squashed together in  single commit (and since classes.xml was split in multiple files, it needs to be rebased to solve that conflict too).

Still couldn't get @reduz to comment on the signal addition, but it doesn't seem harmful, and the use case is cool :)

l1ebl committed:
Add polygon_changed signal to Polygon2D

poke1024 commented:
I've seen in the meantime that there a frequent calls of `_change_notify("some_property")` in Godot's editor code (for example in `Sprite::set_offset`); that might be the pattern one wants to use here too maybe, but as far as I can tell `_change_notify` is purely editor-internal and not callable/interfacable from gdscript. Furthermore, maybe, this PR's signal might even be useful for non-editor runtime code.

reduz commented:
As I mentioned before, Godot does not emit signals on data set functions unless dealing with a shared Resource. This PR is inconsistent with the rest of the engine. I understand what you are trying to do, but we should look for another way to do it, either adding a Polygon resource or something else.

reduz commented:
I will close this, because it's inconsistent. If you want this behavior, you should look into making the polygon inside Polygon2D a shared resource.
